From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sat, 25 Jun 2022 17:21:57 +0200
Subject: [PATCH 1/1] [WasmFS] Fix FS.readFile + WASM_BIGINT

`_wasmfs_read_file` wrote a 32-bit unsigned integer, while
`makeGetValue` in JS read this as a 64-bit integer.

This caused the length to be returned incorrectly when linking
with `-sWASM_BIGINT`. Fix this by ensuring that `_wasmfs_read_file`
will write a 64-bit integer.

Upstream-Status: Accepted [https://github.com/emscripten-core/emscripten/commit/e79f10ebdbd8649324bfd84933548bfee55d566a]

diff --git a/src/library_wasmfs.js b/src/library_wasmfs.js
index 1111111..2222222 100644
--- a/src/library_wasmfs.js
+++ b/src/library_wasmfs.js
@@ -18,6 +18,7 @@ mergeInto(LibraryManager.library, {
     '$PATH',
     '$allocateUTF8',
     '$allocateUTF8OnStack',
+    '$readI53FromI64',
   ],
   $FS : {
     // TODO: Clean up the following functions - currently copied from library_fs.js directly.
@@ -88,8 +89,8 @@ mergeInto(LibraryManager.library, {
 
       // Copy the file into a JS buffer on the heap.
       var buf = __wasmfs_read_file(pathName);
-      // The integer length is returned in the first 8 bytes of the buffer.
-      var length = {{{ makeGetValue('buf', '0', 'i64') }}};
+      // The signed integer length resides in the first 8 bytes of the buffer.
+      var length = {{{ makeGetValue('buf', '0', 'i53') }}};
 
       // Default return type is binary.
       // The buffer contents exist 8 bytes after the returned pointer.
diff --git a/src/parseTools.js b/src/parseTools.js
index 1111111..2222222 100644
--- a/src/parseTools.js
+++ b/src/parseTools.js
@@ -408,6 +408,9 @@ function makeGetValue(ptr, pos, type, noNeedFirst, unsigned, ignore, align) {
   }
 
   const offset = calcFastOffset(ptr, pos, noNeedFirst);
+  if (type === 'i53' || type === 'u53') {
+    return 'readI53From' + (unsigned ? 'U' : 'I') + '64(' + offset + ')';
+  }
 
   const slab = getHeapForType(type);
   let ret = slab + '[' + getHeapOffset(offset, type) + ']';
diff --git a/system/lib/wasmfs/js_api.cpp b/system/lib/wasmfs/js_api.cpp
index 1111111..2222222 100644
--- a/system/lib/wasmfs/js_api.cpp
+++ b/system/lib/wasmfs/js_api.cpp
@@ -22,6 +22,8 @@ __wasi_fd_t wasmfs_create_file(char* pathname, mode_t mode, backend_t backend);
 // The buffer will also contain the file length.
 // Caller must free the returned pointer.
 void* _wasmfs_read_file(char* path) {
+  static_assert(sizeof(off_t) == 8, "File offset type must be 64-bit");
+
   struct stat file;
   int err = 0;
   err = stat(path, &file);
@@ -34,8 +36,8 @@ void* _wasmfs_read_file(char* path) {
   // first 8 bytes. The remaining bytes will contain the buffer contents. This
   // allows the caller to use HEAPU8.subarray(buf + 8, buf + 8 + length).
   off_t size = file.st_size;
-  uint8_t* result = (uint8_t*)malloc((size + sizeof(size)));
-  *(uint32_t*)result = size;
+  uint8_t* result = (uint8_t*)malloc(size + sizeof(size));
+  *(off_t*)result = size;
 
   int fd = open(path, O_RDONLY);
   if (fd < 0) {
